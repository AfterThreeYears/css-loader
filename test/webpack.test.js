import fs from 'fs';
import path from 'path';
import webpack from 'webpack';

function mkdir(p) {
  try {
    fs.mkdirSync(p);
  } catch (e) {
    // ignore
  }
}

describe('webpack', () => {
  const fixtures = path.resolve(__dirname, 'webpack-fixtures');
  const output = path.resolve(__dirname, '.output');
  const tests = fs.readdirSync(fixtures);
  tests.forEach((filename) => {
    it(`should process ${filename}`, () => new Promise((resolve, reject) => {
      const outputDirectory = path.resolve(output, filename.replace(/[^a-zA-Z0-9]+/g, '_'));
      mkdir(outputDirectory);
      webpack({
        entry: `./${filename}/index.css`,
        context: fixtures,
        output: {
          path: outputDirectory,
          filename: 'bundle.js',
          libraryTarget: 'commonjs2',
        },
        module: {
          rules: [
            {
              test: /\.css$/,
              use: {
                loader: path.resolve(__dirname, '../src'),
                options: {
                  sourceMap: true,
                  sourceMapContext: __dirname,
                },
              },
            },
          ],
        },
      }, (err, stats) => {
        if (err) {
          reject(err);
          return;
        }
        try {
          expect(stats.toString('errors-only')).toMatchSnapshot();
          const result = require(path.resolve(outputDirectory, 'bundle.js')); // eslint-disable-line
          expect(result).toMatchSnapshot();
          expect(result.default.toString()).toMatchSnapshot();
          for (const item of result.default) {
            fs.writeFileSync(path.resolve(outputDirectory, `result${item[0]}.css`), item[1], 'utf-8');
            item[3].file = `result${item[0]}.css`;
            fs.writeFileSync(path.resolve(outputDirectory, `result${item[0]}.css.map`), JSON.stringify(item[3]), 'utf-8');
          }
        } catch (e) {
          reject(e);
          return;
        }
        resolve();
      });
    }));
  });
});
