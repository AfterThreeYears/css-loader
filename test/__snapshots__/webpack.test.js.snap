// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack should process at-import 1`] = `""`;

exports[`webpack should process at-import 2`] = `
Object {
  "default": Array [
    Array [
      1,
      "body {
  background: blue;
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAAA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/at-import/other.css",
        ],
        "sourcesContent": Array [
          "body {
  background: blue;
}
",
        ],
        "version": 3,
      },
    ],
    Array [
      2,
      "body {
  background: green;
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAEA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/at-import/abc.css",
        ],
        "sourcesContent": Array [
          "@import './other.css';

body {
  background: green;
}
",
        ],
        "version": 3,
      },
    ],
    Array [
      3,
      "body {
  background: red;
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAGA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/at-import/index.css",
        ],
        "sourcesContent": Array [
          "@import './abc.css';
@import './other.css';

body {
  background: red;
}
",
        ],
        "version": 3,
      },
    ],
  ],
}
`;

exports[`webpack should process at-import 3`] = `
"body {
  background: blue;
}

body {
  background: green;
}

body {
  background: red;
}
"
`;

exports[`webpack should process at-import-media 1`] = `""`;

exports[`webpack should process at-import-media 2`] = `
Object {
  "default": Array [
    Array [
      3,
      "body {
  background: blue;
}
",
      "(screen) and (print)",
      Object {
        "file": "x",
        "mappings": "AAAA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/at-import-media/other.css",
        ],
        "sourcesContent": Array [
          "body {
  background: blue;
}
",
        ],
        "version": 3,
      },
    ],
    Array [
      1,
      "body {
  background: green;
}
",
      "print",
      Object {
        "file": "x",
        "mappings": "AAEA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/at-import-media/abc.css",
        ],
        "sourcesContent": Array [
          "@import './other.css' screen;

body {
  background: green;
}
",
        ],
        "version": 3,
      },
    ],
    Array [
      2,
      "body {
  background: red;
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAEA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/at-import-media/index.css",
        ],
        "sourcesContent": Array [
          "@import './abc.css' print;

body {
  background: red;
}
",
        ],
        "version": 3,
      },
    ],
  ],
}
`;

exports[`webpack should process at-import-media 3`] = `
"@media (screen) and (print){body {
  background: blue;
}
}
@media print{body {
  background: green;
}
}
body {
  background: red;
}
"
`;

exports[`webpack should process import-export 1`] = `""`;

exports[`webpack should process import-export 2`] = `
Object {
  "default": Array [
    Array [
      1,
      "body {
  background: red;
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAIA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/import-export/abc.css",
        ],
        "sourcesContent": Array [
          ":export {
  def: 123px;
}

body {
  background: red;
}
",
        ],
        "version": 3,
      },
    ],
    Array [
      2,
      "body {
  background: green;
}

",
      "",
      Object {
        "file": "x",
        "mappings": "AAAA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/import-export/other.css",
        ],
        "sourcesContent": Array [
          "body {
  background: green;
}

:export {
  abc: red;
  def: blue;
}
",
        ],
        "version": 3,
      },
    ],
    Array [
      3,
      "body {
  color: red;
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAqBA;AACA;AACA;",
        "sources": Array [
          "webpack-css:///./webpack-fixtures/import-export/index.css",
        ],
        "sourcesContent": Array [
          "@import url(\\"./abc.css\\");

:import(\\"./other.css\\") {
  abc: abc;
}

:import(\\"./abc.css\\") {
  my-import: def;
}

:export {
  xyz1: abc;
  xyz2: red, 1px;
  xyz3: my-import, 1px;
  xyz4: no;
}

:export {
  xyz4: yes;
}

body {
  color: abc;
}
",
        ],
        "version": 3,
      },
    ],
  ],
  "xyz1": "red",
  "xyz2": "red, 1px",
  "xyz3": "123px, 1px",
  "xyz4": "yes",
}
`;

exports[`webpack should process import-export 3`] = `
"body {
  background: red;
}

body {
  background: green;
}


body {
  color: red;
}
"
`;

exports[`webpack should process simple 1`] = `""`;

exports[`webpack should process simple 2`] = `
Object {
  "default": Array [
    Array [
      1,
      "body {
  background: red;
}

.class {
  color: rgb(1, 2, 3);
}
",
      "",
      Object {
        "file": "x",
        "mappings": "AAAA;AACA;AACA;;AAEA;AACA;AACA",
        "names": Array [],
        "sources": Array [
          "webpack-css:///./webpack-fixtures/simple/index.css",
        ],
        "sourcesContent": Array [
          "body {
  background: red;
}

.class {
  color: rgb(1, 2, 3);
}
",
        ],
        "version": 3,
      },
    ],
  ],
}
`;

exports[`webpack should process simple 3`] = `
"body {
  background: red;
}

.class {
  color: rgb(1, 2, 3);
}
"
`;
